modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/NetworkDirectionProcessor.java

```diff
@@ -227,31 +246,53 @@ public String getType() {
     }
 
     public static final class Factory implements Processor.Factory {
-
+        private final ScriptService scriptService;
         static final String DEFAULT_SOURCE_IP = "source.ip";
         static final String DEFAULT_DEST_IP = "destination.ip";
         static final String DEFAULT_TARGET = "network.direction";
 
+        public Factory(ScriptService scriptService) {
+            this.scriptService = scriptService;
+        }
+
         @Override
         public NetworkDirectionProcessor create(
             Map<String, Processor.Factory> registry,
             String processorTag,
             String description,
             Map<String, Object> config
         ) throws Exception {
-            String sourceIpField = ConfigurationUtils.readStringProperty(TYPE, processorTag, config, "source_ip", DEFAULT_SOURCE_IP);
-            String destIpField = ConfigurationUtils.readStringProperty(TYPE, processorTag, config, "destination_ip", DEFAULT_DEST_IP);
-            String targetField = ConfigurationUtils.readStringProperty(TYPE, processorTag, config, "target_field", DEFAULT_TARGET);
-            List<String> internalNetworks = ConfigurationUtils.readList(TYPE, processorTag, config, "internal_networks");
-            boolean ignoreMissing = readBooleanProperty(TYPE, processorTag, config, "ignore_missing", true);
+            final String sourceIpField = ConfigurationUtils.readStringProperty(TYPE, processorTag, config, "source_ip", DEFAULT_SOURCE_IP);
+            final String destIpField = ConfigurationUtils.readStringProperty(TYPE, processorTag, config, "destination_ip", DEFAULT_DEST_IP);
+            final String targetField = ConfigurationUtils.readStringProperty(TYPE, processorTag, config, "target_field", DEFAULT_TARGET);
+            final boolean ignoreMissing = readBooleanProperty(TYPE, processorTag, config, "ignore_missing", true);
+
+            final List<String> internalNetworks = ConfigurationUtils.readOptionalList(TYPE, processorTag, config, "internal_networks");
+            final String internalNetworksField = ConfigurationUtils.readOptionalStringProperty(
+                TYPE,
+                processorTag,
+                config,
+                "internal_networks_field"
+            );
```

Comment:
I believe you can keep just the `internal_networks` field and have it support either an explicit list of values or a template snippet. See the use of `ValueSource` in `AppendProcessor` as an example.

Reply:
If the `internal_networks_field` option can be eliminated, obviously that will imply a number of other simplifications throughout this PR.

Reply:
So, I took a pass at doing this, and maybe it's my misunderstanding of the mustache-templated fields, but I need to support referencing something like this:

```
curl -H "Content-Type: application/json" -X POST \
-u elastic:password http://localhost:9200/_ingest/pipeline/_simulate\?verbose \
--data-binary @- << EOF
{
  "pipeline": {
    "processors": [
      {
        "network_direction": {
          "internal_networks": "{{test}}"
        }
      }
    ]
  },
  "docs": [
    {
      "_source": {
        "test": ["public"],
        "source": {
          "ip": "128.232.110.120"
        },
        "destination": {
          "ip": "192.168.1.1"
        }
      }
    }
  ]
}
EOF
```

When resolving the templated value, I get ```{0=public}``` as the resulting `internal_networks` value. Is it possible to for a field template to reference a field that contains an array and for it to be kept as an array and not turned into a string? Because if not I think we might need to keep the additional field.

Reply:
Thanks for the explanation. Templates won't present those values as arrays, so I think we'll need to keep the two separate configuration options that you have here.

Reply:
great, I'll update the docs and make the other requested changes, thanks for the review